src/core/qgssqlstatement.cpp

```diff
@@ -136,33 +136,33 @@ QgsSQLStatement::QgsSQLStatement( const QString &expr )
 QgsSQLStatement::QgsSQLStatement( const QString &expr, bool allowFragments )
   : mAllowFragments( allowFragments )
 {
-  mRootNode = ::parse( expr, mParserErrorString, mAllowFragments );
+  mRootNode.reset( ::parse( expr, mParserErrorString, mAllowFragments ) );
   mStatement = expr;
 }
 
 QgsSQLStatement::QgsSQLStatement( const QgsSQLStatement &other )
   : mAllowFragments( other.mAllowFragments )
   , mStatement( other.mStatement )
 {
-  mRootNode = ::parse( mStatement, mParserErrorString, mAllowFragments );
+  mRootNode.reset( ::parse( mStatement, mParserErrorString, mAllowFragments ) );
 }
 
 QgsSQLStatement &QgsSQLStatement::operator=( const QgsSQLStatement &other )
 {
   if ( &other != this )
   {
-    delete mRootNode;
+    mRootNode.reset( ::parse( mStatement, mParserErrorString, mAllowFragments ) );
     mParserErrorString.clear();
     mStatement = other.mStatement;
     mAllowFragments = other.mAllowFragments;
```

Comment:
The order of these need to be swapped -- you're using the wrong values for mStatement,, mAllowFragment when parsing otherwise.

Reply:
Done.